lib/Alchemy/Phrasea/Authentication/Provider/Phraseanet.php

```diff
@@ -0,0 +1,320 @@
+<?php
+
+/*
+ * This file is part of Phraseanet
+ *
+ * (c) 2005-2016 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Authentication\Provider;
+
+use Alchemy\Phrasea\Authentication\Exception\NotAuthenticatedException;
+use Alchemy\Phrasea\Authentication\Provider\Token\Identity;
+use Alchemy\Phrasea\Authentication\Provider\Token\Token;
+use Alchemy\Phrasea\Exception\InvalidArgumentException;
+use Guzzle\Common\Exception\GuzzleException;
+use Guzzle\Http\Client as Guzzle;
+use Guzzle\Http\ClientInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Routing\Generator\UrlGenerator;
+
+
+class Phraseanet extends AbstractProvider
+{
+    private $baseurl;
+    private $key;
+    private $secret;
+
+    private $client;
+
+    public function __construct(UrlGenerator $generator, SessionInterface $session, $id, $display, $title, array $options, Guzzle $client)
+    {
+        parent::__construct($generator, $session, $id, $display, $title);
+
+        $this->baseurl = $options['base-url'];
+        $this->key     = $options['client-id'];
+        $this->secret  = $options['client-secret'];
+
+        $this->client  = $client;
+    }
+
+    /**
+     * @param ClientInterface $client
+     *
+     * @return Phraseanet
+     */
+    public function setGuzzleClient(ClientInterface $client)
+    {
+        $this->client = $client;
+
+        return $this;
+    }
+
+    /**
+     * @return ClientInterface
+     */
+    public function getGuzzleClient()
+    {
+        return $this->client;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function authenticate(array $params = array())
+    {
+        $params = array_merge(['providerId' => $this->getId()], $params);
+
+        $state = $this->createState();
+
+        $this->session->set($this->id . '.provider.state', $state);
+
+        return new RedirectResponse($this->baseurl . '/login/oauth/authorize?' . http_build_query([
```

Comment:
Actually the authorize URL looks like this:
`{baseURL}/{providerType}/{providerName}/authorize`

where:
- `baseURL` is the server root URI.
- `providerType` is either `oauth` or `saml`
- `providerName` is the authentication provider behind Auth service.

> i.e. https://ps-auth.com/oauth/facebook/authorize

So we can set `https://ps-auth.com/oauth/facebook` as baseURL and have:
```suggestion
        return new RedirectResponse($this->baseurl . '/authorize?' . http_build_query([
```

or we can introduce `providerType` and `providerName` option parameters:
```suggestion
        return new RedirectResponse(sprintf('%s/%s/%s/authorize?%s', $this->baseurl, $this->providerType, $this->providerName, http_build_query([
```

Reply:
The provider "type" is the actual classname passed to the factory. I renamed to "PhraseanetOauth" provider, so we don't really need a kind of "subtype" in conf (providerType oauth|saml). Since Oauth and Saml is quite different code, I suggest it's easier to have 2 separated implementations, even if some url contains hardcoded "oauth" or "saml" parts. 
But I agree that adding a variable part "providerName" from conf/options allows to have many (same type) providers on same baseurl.